src/compute/src/sink/persist_sink.rs

```diff
@@ -1105,12 +1125,66 @@ where
                         }
                     }
                 }
+                _it_changed = read_only.changed() => {
+                    // We might have to append some batches that we stashed
+                    // while in read-only mode.
+                }
                 else => {
                     // All inputs are exhausted, so we can shut down.
                     return;
                 }
             };
 
+            // Only retain descriptions and batches that still have a chance of
+            // being applied.
+            let persist_upper = write.shared_upper();
+            in_flight_descriptions.retain(|(_lower, upper)| {
+                PartialOrder::less_equal(&persist_upper, upper)
+            });
+
+            for ((_lower, upper), batch_set) in in_flight_batches.iter_mut() {
+                if PartialOrder::less_equal(&persist_upper, upper) {
+                    continue;
+                }
+
+                // We're not keeping this batch. Be nice and delete any data
+                // that has been written.
+                for batch in batch_set.finished.drain(..) {
+                    batch.delete().await;
+                }
+                if let Some(batch) = batch_set.incomplete.take() {
+                    batch
+                        .finish(upper.clone()).await
+                        .expect("invalid usage")
+                        .delete().await;
+                }
+            }
+            // WIP: It's annoying that we're first iterating and doing the
+            // retain, but we can't do the batch deletion inside retain because
+            // we need async.
```

Comment:
Should this still be a "WIP" comment?

Reply:
It's still annoying but I didn't find a better solution. Will remove the `WIP` part